Redesign Search page with voice search

candry7731 opened the pull request:
Add Voice search while in folders 
Also added a spinner to grid items when loading so user know the app is working
This is to resolve #587

neilsb commented:
Is this PR intended to replace the Search Redesign (#568) or should this be looked at after 568 has been merged?

jimdogx commented:
This is looking really good!!!!

A couple of things:

* Voice search appears to work everywhere for me except Live TV.  If I try to voice search, it doesn't find anything, but the bad part is that you lose focus from the TV guide and can't get it back.  Might be worth just seeing if you can remove voice search from the TV guide all together.  Because it works great for Movies, TVShows, etc.

* Roku is spitting out a few warnings / errors (even though it appears to work).  Might be worth investigating:

1. 
*** ERROR: Missing or invalid PHY: '/RokuOS/Keyboards/MiniKeyboardKDF.json'

2.
Warning occurred while setting a field of an RoSGNode
-- Type mismatch occurred when setting the "results" field of a "SearchTask" node
pkg:/components/search/SearchTask.brs(7)

3.
BRIGHTSCRIPT: ERROR: roUrlTransfer.GetToString: Sync-wait interrupted: pkg:/source/api/baserequest.brs(71)

jimdogx commented on components/ItemGrid/LoadItemsTask2.brs:
```diff
@@ -85,5 +90,5 @@ sub loadItems()
 
     m.top.content = results
-
+    print "Params" params
```
Don't forget to remove debug statements 😄

jimdogx commented on components/liveTv/schedule.brs:
```diff
@@ -183,53 +165,6 @@ sub onWatchChannelSelected()
     m.top.watchChannel = m.detailsPane.channel
 end sub
 
```
I might be missing something, but why was the record functionality removed?  I don't see it replaced anywhere else.

candry7731 replied:
It wasn't supposed to be removed..... can you add it back?

candry7731 commented:
I wonder why live tv doesn't work..... we can hide for now, ill look into
this more.


1. Roku is looking for a custom keyboard file, however I think k the
default one is fine....

2. Not sure about this one...

3. The api request is started and stopped everytime you type a letter or
say a letter.... thats the error it throws.... I don't think this is an
issue....


Did you notice/test the server and username/password keyboard was also
updated to the voice keyboard?


On Sat, May 14, 2022, 5:35 PM Jimi ***@***.***> wrote:

> This is looking really good!!!!
>
> A couple of things:
>
>    -
>
>    Voice search appears to work everywhere for me except Live TV. If I
>    try to voice search, it doesn't find anything, but the bad part is that you
>    lose focus from the TV guide and can't get it back. Might be worth just
>    seeing if you can remove voice search from the TV guide all together.
>    Because it works great for Movies, TVShows, etc.
>    -
>
>    Roku is spitting out a few warnings / errors (even though it appears
>    to work). Might be worth investigating:
>
>
>    1.
>
> *** ERROR: Missing or invalid PHY: '/RokuOS/Keyboards/MiniKeyboardKDF.json'
>
>    1.
>
> Warning occurred while setting a field of an RoSGNode
> -- Type mismatch occurred when setting the "results" field of a
> "SearchTask" node
> pkg:/components/search/SearchTask.brs(7)
>
>    1.
>
> BRIGHTSCRIPT: ERROR: roUrlTransfer.GetToString: Sync-wait interrupted:
> pkg:/source/api/baserequest.brs(71)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin-roku/pull/593#issuecomment-1126821623>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT6PDRAKHK3C23FNMK4OXETVKATDDANCNFSM5VS2W6CQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

candry7731 commented:
Working on voice functions for channels now...

1hitsong commented:
Is there a size limit on the search phrase? 

I said "a nightmare on elm street" but it searched for "a nightmare on"

candry7731 commented:
> 1. Is there a size limit on the search phrase? _This has been increased to  max 75 characters _

> 2. Can we add the ability to modify the search? _The voice search at the top is intended to be a quick voice search, if you want to modify the text you can click the options -> search and see the voice search with the full keyboard. If you use an external remote like the roku app on your phone you can remove the space and type into that field_

candry7731 commented:
This should be finished and ready for merge...

candry7731 commented:
This has been fixed, can you test?

On Sat, May 21, 2022, 4:23 PM 1hitsong ***@***.***> wrote:

> Is there a size limit on the search phrase?
>
> I said "a nightmare on elm street" but it searched for "a nightmare on"
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin-roku/pull/593#issuecomment-1133770207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT6PDRACCH52MJDZCXTWYO3VLFH5NANCNFSM5VS2W6CQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jimdogx commented:
Appears to work everywhere except for "Collections".

jimdogx commented on components/liveTv/schedule.brs:
```diff
@@ -125,17 +135,17 @@ end sub
 sub onProgramDetailsLoaded()
-    if m.LoadProgramDetailsTask.programDetails = invalid then return
-    channel = m.scheduleGrid.content.GetChild(m.LoadProgramDetailsTask.programDetails.channelIndex)
+    if m.LoadProgramDetailsTask.programDetails = invalid
```
Removing "then return" breaks live tv recording.  Same for removing m.scheduleGrid.showLoadingDataFeedback = false below.

candry7731 commented:
1. I have moved the Search Microphone to be above the alpha picker.
2. if you press the replay button (Backwards round arrow) on the remote it clears all search filters and returns all content
3. Updated the overhang title to update on alpha picker selection or voice search selection so user knows what is being filtered

candry7731 commented:
@neilsb Please check these updates. I think we are close :) I updated everything you mentioned in your comments... I think it works/looks a lot better. let me know

candry7731 commented:
@neilsb Where are we with this?

1hitsong commented:
I pressed the mic button, but didn't say anything. This results in a blank search box. Since the search is now on the right, I expected pressing left would bring the movies back into focus, but it was "down" that refocused it.

When I voice search in collections, the title shows "(filtered by...)" but it didn't filter the collections. They all still appear.

candry7731 commented:
Ahh, collections.... I will look at this tonight.

Thanks!

On Fri, May 27, 2022, 4:28 PM 1hitsong ***@***.***> wrote:

> I pressed the mic button, but didn't say anything. This results in a blank
> search box. Since the search is now on the right, I expected pressing left
> would bring the movies back into focus, but it was "down" that refocused it.
>
> When I voice search in collections, the title shows "(filtered by...)" but
> it didn't filter the collections. They all still appear.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin-roku/pull/593#issuecomment-1140052489>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT6PDREJFY6O74PYT7WMWVDVME47VANCNFSM5VS2W6CQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

cewert commented on source/Main.brs:
```diff
@@ -187,27 +187,29 @@ sub Main (args as dynamic) as void
-            dialog = createObject("roSGNode", "ProgressDialog")
-            dialog.title = tr("Loading Search Data")
-            m.scene.dialog = dialog
+            'options.setFocus(true)
+            group.findNode("SearchSpinner").visible = true
+            'dialog = createObject("roSGNode", "ProgressDialog")
+            'dialog.title = tr("Loading Search Data")
+            'm.scene.dialog = dialog
```
```suggestion
            searchSpinner = group.findNode("SearchSpinner")
            searchSpinner.visible = true
```

cewert replied:
If you ever find yourself searching for the same node more than once, best practice is to save the node reference to a variable and use that instead. It's more efficient that way.

cewert commented on source/Main.brs:
```diff
@@ -187,27 +187,29 @@ sub Main (args as dynamic) as void
+            'm.scene.dialog = dialog
             results = SearchMedia(query)
-            dialog.close = true
+            group.findNode("SearchSpinner").visible = false
```
```suggestion
            searchSpinner.visible = false
```

cewert commented on source/api/Items.brs:
```diff
@@ -38,30 +38,42 @@ function SearchMedia(query as string)
-    end for
-    data.SearchHints = results
-    return data
+    if query = ""
+
+    else
+        resp = APIRequest(Substitute("Search/Hints", get_setting("active_user")), {
+            "searchTerm": query,
+            "IncludePeople": true,
+            "IncludeMedia": true,
+            "IncludeShows": true,
+            "IncludeGenres": true,
+            "IncludeStudios": true,
+            "IncludeArtists": true,
+            "IncludeItemTypes": "LiveTvChannel,Movie,BoxSet,Series,Episode,Video,Person"
+            "EnableTotalRecordCount": false,
+            "ImageTypeLimit": 1,
+            "Recursive": true,
+            "limit": 100
+        })
+
+        ' TODO/FIXME:
+        ' Notice the Limit = 25.  My first attempt returned over 1000 results.
+        ' Even at 25 it takes a good 3 or 4 seconds to return... so you may have
+        ' to pause data entry or something while results come back...
+
+        data = getJson(resp)
+        results = []
+        for each item in data.SearchHints
+            tmp = CreateObject("roSGNode", "SearchData")
+            tmp.image = PosterImage(item.id)
+            tmp.json = item
+            results.push(tmp)
+        end for
+        data.SearchHints = results
+        return data
+    end if
```
```suggestion
    if query <> ""
        resp = APIRequest(Substitute("Search/Hints", get_setting("active_user")), {
            "searchTerm": query,
            "IncludePeople": true,
            "IncludeMedia": true,
            "IncludeShows": true,
            "IncludeGenres": true,
            "IncludeStudios": true,
            "IncludeArtists": true,
            "IncludeItemTypes": "LiveTvChannel,Movie,BoxSet,Series,Episode,Video,Person"
            "EnableTotalRecordCount": false,
            "ImageTypeLimit": 1,
            "Recursive": true,
            "limit": 100
        })

        ' TODO/FIXME:
        ' Notice the Limit = 25.  My first attempt returned over 1000 results.
        ' Even at 25 it takes a good 3 or 4 seconds to return... so you may have
        ' to pause data entry or something while results come back...

        data = getJson(resp)
        results = []
        for each item in data.SearchHints
            tmp = CreateObject("roSGNode", "SearchData")
            tmp.image = PosterImage(item.id)
            tmp.json = item
            results.push(tmp)
        end for
        data.SearchHints = results
        return data
    end if
```

cewert commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -26,6 +26,16 @@ sub init()
     m.itemGrid.observeField("AlphaSelected", "onItemAlphaSelected")
+
+    'Voice filter setup
+    m.VoiceBox = m.top.findNode("VoiceBox")
```
Variables should be lowerCamelCase. Most of your code is but looks like this one snuck by
```suggestion
    m.voiceBox = m.top.findNode("VoiceBox")
```

cewert reviewed (changes requested):
Great job on everything and thanks for joining the dark side and writing brightscript with us 😄

candry7731 commented:
Updated to work with collections

Know issues:
1. When using voice then using the alpha picker you have to select the letter 2 times... Can't figure out why....
2. The microphone icon disappears when you are inside a collection folder... Can't figure out why....

Let me know what you guys think :)

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -63,8 +75,12 @@ sub loadInitialItems()
             m.view = "tvGuide"
+            'Move mic to be visiable on TV Guide screen
+            m.VoiceBox.translation = "[1845, 90]"
+            print "View Setting: " viewSetting
```
```suggestion
```

Remove debug printing

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -63,8 +75,12 @@ sub loadInitialItems()
+            print "View Setting: " viewSetting
         else
             m.view = "livetv"
+            'm.VoiceBox.visible = "false"
```
```suggestion
```

Remove commented code

neilsb commented on components/liveTv/schedule.brs:
```diff
@@ -42,6 +39,19 @@ sub channelFilterSet()
 
+'Voice Search set
+sub channelsearchTermSet()
+    print "Channel Filter set"
```
```suggestion
```
Remove debug statement

neilsb commented on source/Main.brs:
```diff
@@ -273,8 +272,8 @@ sub Main (args as dynamic) as void
                 sceneManager.callFunc("pushScene", group)
-                group.findNode("SearchBox").findNode("search-input").setFocus(true)
-                group.findNode("SearchBox").findNode("search-input").active = true
+                group.findNode("SearchBox").findNode("search_Key").setFocus(true)
+                group.findNode("SearchBox").findNode("search_Key").active = true
```
```suggestion
	        searchKey = group.findNode("SearchBox").findNode("search_Key")
                searchKey.setFocus(true)
                searchKey.active = true
```

Slightly more efficient to save the reference to SearchBox rather than calling `findNode` 4 times.

neilsb reviewed (changes requested):
Looking good.  A couple of minor suggestions but other than that I think this is OK to go as a big improvement to the search.

cewert commented on source/Main.brs:
```diff
@@ -273,8 +272,9 @@ sub Main (args as dynamic) as void
                 sceneManager.callFunc("pushScene", group)
-                group.findNode("SearchBox").findNode("search-input").setFocus(true)
-                group.findNode("SearchBox").findNode("search-input").active = true
+	        searchKey = group.findNode("SearchBox").findNode("search_Key")
```
need to make CI happy

```suggestion
	            searchKey = group.findNode("SearchBox").findNode("search_Key")
```

neilsb commented:
The voice search the libraries isn't working correctly any more.  The search triggers fine, but instead of filtering the results (as it did previously) it now just displays a white area on the right hand side of the screen.  When this happens, the Arrow and OK buttons don't do anything.  Pressing "Replay" closes the white back and returns you to the item grid, and the "Back" button takes you back to the home screen.

![image](https://user-images.githubusercontent.com/5013482/176140456-653cd811-9c97-4b69-9ae6-8cf51e378c4a.png)

candry7731 commented:
This has been updated and tested.

neilsb commented:
Looking good to me now.   Just need @cewert  and @jimdogx  to approve (since they have a reviews marked as requested changes)

jimdogx commented:
> Looking good to me now. Just need @cewert and @jimdogx to approve (since they have a reviews marked as requested changes)

I will test this weekend 👍

candry7731 commented:
Thanks! What happens if @cewert doesn't approve his requested changes? It
doesn't seem like he is as active as you guys...


On Fri, Jul 1, 2022 at 12:20 PM Jimi ***@***.***> wrote:

> Looking good to me now. Just need @cewert <https://github.com/cewert> and
> @jimdogx <https://github.com/jimdogx> to approve (since they have a
> reviews marked as requested changes)
>
> I will test this weekend 👍
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin-roku/pull/593#issuecomment-1172560689>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT6PDRATT2T37DQOIU34YZTVR4SD7ANCNFSM5VS2W6CQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jimdogx commented:
1. All of the option translations are not working, you only get placeholder text (NO_ITEMS, TAB_ORDER, etc).  I seem to remember seeing that mentioned by @neilsb somewhere else... so maybe that's not part of this PR?

2. Still does not work for collections.

3. Home screen search keyboard says "...arch Query" in entry box.

4. Using the home screen search, it crashes if I search for a music artist (ListPoster.brs; line 76; userdata is invalid):
```
       if not itemData.json.userdata.played
``` 

Tagging @1hitsong in case he wants to look in to \# 4.

1hitsong commented:
> Using the home screen search, it crashes if I search for a music artist (ListPoster.brs; line 76; userdata is invalid):

Can you send me a video or steps with photos? I'm unable to reproduce this based on what I believe you're instructing.

1hitsong commented:
I'm finding all my music artists are doubled in search results. Clicking on one leads to a dead end, while clicking on the 2nd leads to their artist page.
![WIN_20220702_17_30_23_Pro](https://user-images.githubusercontent.com/3330318/177016743-8f15767f-d461-4189-9c72-0a02ec37201b.jpg)

![WIN_20220702_17_30_37_Pro](https://user-images.githubusercontent.com/3330318/177016746-9c9c072e-e138-4b89-9ae9-7c7aae9d5e7d.jpg)

![WIN_20220702_17_30_42_Pro](https://user-images.githubusercontent.com/3330318/177016747-245aa91f-6120-42c3-bce4-9b51f93a9bd2.jpg)

1hitsong commented:
Search box still says "...arch Query"

![WIN_20220702_17_35_15_Pro](https://user-images.githubusercontent.com/3330318/177016786-ced5d90e-8fa7-4d95-9942-63d22ffe57ab.jpg)

candry7731 commented:
> I'm finding all my music artists are doubled in search results. Clicking on one leads to a dead end, while clicking on the 2nd leads to their artist page. ![WIN_20220702_17_30_23_Pro](https://user-images.githubusercontent.com/3330318/177016743-8f15767f-d461-4189-9c72-0a02ec37201b.jpg)
> 
> ![WIN_20220702_17_30_37_Pro](https://user-images.githubusercontent.com/3330318/177016746-9c9c072e-e138-4b89-9ae9-7c7aae9d5e7d.jpg)
> 
> ![WIN_20220702_17_30_42_Pro](https://user-images.githubusercontent.com/3330318/177016747-245aa91f-6120-42c3-bce4-9b51f93a9bd2.jpg)

I do see the doubling artist but I can click on either one and see the albums and songs associated with the artist.

candry7731 commented:
@jimdogx I am not getting any errors when searching an artist.

jimdogx commented:
Turns out the music artist I searched for is also named in two TV episodes....and the crash was when trying to blur the episodes.

jimdogx commented on source/api/Items.brs:
```diff
@@ -34,34 +34,44 @@ function ItemPostPlaybackInfo(id as string, mediaSourceId = "" as string, audioT
+            "limit": 100
+        })
+
+        ' TODO/FIXME:
```
You can remove my TODO/FIXME comment from a while ago when there wasn't a limit.  100 seems to be working fine.

jimdogx commented on source/ShowScenes.brs:
```diff
@@ -438,12 +438,16 @@ end function
 
-    search = group.findNode("SearchBox")
-    search.observeField("search_value", m.port)
+    'search = group.findNode("searchBox")
```
This block of commented code with the TODO/FIXME can be deleted since your new "observeField" is the one you want here now that it's fully implemented.

candry7731 commented:
Do I need to resolve the conflict to get this merged? Is there anything else that needs to be fixed/updated?

jimdogx commented:
> 
Do I need to resolve the conflict to get this merged? Is there anything else that needs to be fixed/updated?

Master merged in (think I got it right)...  just need @cewert to give it the final 👍

candry7731 commented:
Thanks Jimi!

On Sat, Jul 9, 2022, 5:02 PM Jimi ***@***.***> wrote:

> Do I need to resolve the conflict to get this merged? Is there anything
> else that needs to be fixed/updated?
>
> Master merged in (think I got it right)... just need @cewert
> <https://github.com/cewert> to give it the final 👍
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin-roku/pull/593#issuecomment-1179612924>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT6PDRCJCUMN6DDPY45DC2TVTHZFRANCNFSM5VS2W6CQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cewert commented on components/ItemGrid/ItemGrid.xml:
```diff
@@ -24,14 +24,16 @@
       itemSpacing = "[ 0, 45 ]" 
       drawFocusFeedback = "false" />  
-      <Label translation="[0,540]" id="emptyText" font="font:LargeSystemFont" width="1920" horizAlign="center" vertAlign="center" height="64" visible="false" />
+      <VoiceTextEditBox id="VoiceBox" visible="true" width = "0" hintText = "" translation = "[1845, 120]" maxTextLength = "75" height="32"/>
+      <Label translation="[0,540]" id="emptyText" font="font:LargeSystemFont" width="1910" horizAlign="center" vertAlign="center" height="64" visible="false" />
```
```suggestion
    <VoiceTextEditBox id="VoiceBox" visible="true" width = "0" hintText = "" translation = "[1845, 120]" maxTextLength = "75" height="32"/>
    <Label translation="[0,540]" id="emptyText" font="font:LargeSystemFont" width="1910" horizAlign="center" vertAlign="center" height="64" visible="false" />
```

cewert commented on components/ItemGrid/ItemGrid.xml:
```diff
@@ -24,14 +24,16 @@
         <FloatFieldInterpolator id = "fadeoutLoaded" key="[0.0, 1.0]" keyValue="[ 0.25, 0.00 ]" fieldToInterp="backdrop.opacity" />
     </Animation>
     <Alpha id="AlphaMenu" />
+    
```
```suggestion
```

cewert commented on components/ItemGrid/LoadItemsTask2.brs:
```diff
@@ -47,9 +51,18 @@ sub loadItems()
         params.append({ IncludeItemTypes: m.top.ItemType })
     end if
 
+    if m.top.searchTerm <> ""
+        m.livetvsearch = m.top.searchTerm
+    else
+        m.livetvsearch = m.top.nameStartsWith
+    end if
+
```
Looks like this if else can be combined with the if statement on line 38

candry7731 replied:
You are correct, this has been fixed

cewert commented on components/config/ConfigList.brs:
```diff
@@ -45,13 +45,18 @@ end function
+    dialog.palette = m.greenPalette
 
     if configField.type = "password"
-        dialog.keyboard.textEditBox.secureMode = true
+        dialog.textEditBox.secureMode = true
```
Was this a bug? I don't understand why voice search would need "password" input fields?

candry7731 replied:
This is used for the username password fields when adding a user profile on the login screens. If you are in a public space or with a group of friends you might not want them to see you enter your password....

cewert replied:
Right I understand what they are and why they are useful but why did you need to change this line of code?

candry7731 replied:
weird, I do not see this change in my current file in VSCode

cewert replied:
FYI: that is the tab we use to review the code. It's how we make comments on the code itself. If you click on the file path right above the code snippet, it takes you to the Files Changed tab as well

cewert commented on components/config/SetServerScreen.brs:
```diff
@@ -107,10 +107,13 @@ sub ScanForServersComplete(event)
     dialog.title = tr("Enter the server name or ip address")
     dialog.buttons = [tr("OK"), tr("Cancel")]
     dialog.text = m.serverUrlTextbox.text
+    m.greenPalette = createObject("roSGNode", "RSGPalette")
+    m.greenPalette.colors = { DialogBackgroundColor: "#2A2B2A" }
+    dialog.palette = m.greenPalette
```
No need to save this node to `m.` since the node is recreated every time the function runs.
```suggestion
    greenPalette = createObject("roSGNode", "RSGPalette")
    greenPalette.colors = { DialogBackgroundColor: "#2A2B2A" }
    dialog.palette = greenPalette
```

cewert commented on components/liveTv/LoadChannelsTask.brs:
```diff
@@ -6,15 +6,41 @@ sub loadChannels()
+    end if
+
     params = {
+        includeItemTypes: "LiveTvChannel",
+        SortBy: sort_field,
+        SortOrder: sort_order,
+        recursive: m.top.recursive,
         UserId: get_setting("active_user")
     }
 
+    ' Handle special case when getting names starting with numeral
+    if m.top.NameStartsWith <> ""
+        if m.top.NameStartsWith = "#"
+            params.NameLessThan = "A"
+        else
+            params.NameStartsWith = m.top.nameStartsWith
+        end if
+    end if
+
+    'Append voice search when there is text
+    if m.top.searchTerm <> ""
+        params.searchTerm = m.top.searchTerm
+    end if
+
     if m.top.filter = "Favorites"
         params.append({ isFavorite: true })
     end if
 
-    url = "LiveTv/Channels"
+    url = Substitute("Users/{0}/Items/", get_setting("active_user"))
```
I still don't have a live TV setup so maybe im missing something but this doesn't seem right to me. `LoadChannelsTask.brs` polling the API for user items instead of livetv channels?

candry7731 replied:
Yes, "LiveTv/Channels" API does not have the following params that are needed for the alpha picker or the search to work,  
searchTerm
nameStartsWith
nameLessThan

cewert commented on components/search/SearchResults.xml:
```diff
@@ -1,13 +1,29 @@
+    <field id="quickPlayNode" type="node" alwaysNotify="true" />
+    <field id="imageDisplayMode" type="string" value="scaleToZoom" />
+    <field id="searchAlpha" type="string" alias="SearchBox.search_values" alwaysNotify="true" onChange="searchMedias" />
+
```
```suggestion
```

cewert commented on components/search/SearchResults.xml:
```diff
@@ -1,13 +1,29 @@
-    <SearchRow id="SearchSelect" visible="false" />
-
-    <OptionsSlider id="options" />
+  <Rectangle width="1920" height="1080" color="#000000" opacity="0.75" />
+    <LayoutGroup layoutDirection="horiz"  id="SearchAlphabox" translation="[70, 120]">
+      <SearchBox id="SearchBox" visible="true" focusable="true"/>
+    </LayoutGroup>
+      <LayoutGroup layoutDirection="vert"  id="searchResults" translation="[470, 150]" >
+        <Label id = "SearchHelpText" text=""/>
+        <SearchRow id="searchSelect" visible="true" focusable="true"/>
+      </LayoutGroup>    
+      <OptionsSlider id="options" />
+        <Spinner id = "searchSpinner"  visible="false" translation="[1050, 500]"/>
```
```suggestion
    <Rectangle width="1920" height="1080" color="#000000" opacity="0.75" />
    <LayoutGroup layoutDirection="horiz"  id="SearchAlphabox" translation="[70, 120]">
      <SearchBox id="SearchBox" visible="true" focusable="true"/>
    </LayoutGroup>
    <LayoutGroup layoutDirection="vert"  id="searchResults" translation="[470, 150]" >
      <Label id = "SearchHelpText" text=""/>
      <SearchRow id="searchSelect" visible="true" focusable="true"/>
    </LayoutGroup>    
    <OptionsSlider id="options" />
    <Spinner id = "searchSpinner"  visible="false" translation="[1050, 500]"/>
```

cewert commented on components/search/SearchRow.brs:
```diff
@@ -11,27 +11,29 @@ sub init()
-    m.top.itemSize = [dimensions["width"] - border * 2, itemHeight]
-    m.top.itemSpacing = [0, 50]
+    m.top.itemSize = [1350, itemHeight] ' this is used for setting the row size
+    ' m.top.itemSize = [dimensions["width"] - border, itemHeight]
```
Is this comment needed?
```suggestion
```

cewert commented on locale/en_US/translations.ts:
```diff
@@ -211,54 +211,54 @@
     <message>
-        <source>IMDB_RATING</source>
+        <source>IMDB_Rating</source>
         <translation>IMDb Rating</translation>
```
```suggestion
        <translation>IMDB Rating</translation>
```

cewert replied:
I was wrong on this, sorry. Just googled it

cewert replied:
I pushed a fix for this

cewert commented:
Done looking at code for now. I'll try to test this tonight

candry7731 commented:
That's is correct!

candry7731 commented:
> Done looking at code for now. I'll try to test this tonight
> 
> Edit: What should I be testing exactly? Looks like search page, voice search in folders, and voice search in collections. is that it? @candry7731 @jimdogx 

That is correct!

cewert commented:
The first thing I tried was using the voice search button on the home screen and I was surprised when I got no local results. Apparently that's a normal roku thing since emby gave me the same results?

One thing I did notice is that on ours it said "No results from dev" as it popped up the roku channel crap. Where on emby it said "no results from emby". Is that something you can change? Or is that even part of this PR? This is my first time using voice search sorry if I'm missing something

cewert reviewed (changes requested):
For the search page, can we add the search query to the overhang? And probably the word search too right? i.e. "Search: searchString" The search query will help fill some of the empty space in general but would be most helpful for voice search since you are more likely to search for a longer string of text. We could make more room in the search box but that takes away room for search results. I think showing the full search query in the overhang is a nice compromise.

cewert commented:
When first landing on the search page it felt "wrong" for lack of better word to go down. I knew A B C D E etc. would be down but it felt wrong since the thing i had selected went down further then the letters i wanted to get to. Once I used it a few times it felt normal however. Maybe we could improve UX a little here by adding some space between the search box and the letters? or you could even get fancy and make it slide down when the search box is selected and slide back up when it's not.

![PXL_20220711_041006613 MP](https://user-images.githubusercontent.com/16514652/178191071-a45dc1d6-caaa-476a-94db-db75f63d0c71.jpg)

cewert commented on components/SearchBox.brs:
```diff
@@ -3,48 +3,25 @@ sub init()
     m.top.vertAlignment = "top"
     m.top.visible = false
+    m.searchText = m.top.findNode("search_Key")
+    m.searchText.textEditBox.hintText = tr("Enter Search Query")
```
I only see "Enter Sear..." which isn't very helpful IMO. Since the box isn't selectable it's also a little confusing too I think. maybe we should change this to something shorter like "Search..."? or even get rid of it entirely and only show the search string?

candry7731 commented:
> When first landing on the search page it felt "wrong" for lack of better word to go down. I knew A B C D E etc. would be down but it felt wrong since the thing i had selected went down further then the letters i wanted to get to. Once I used it a few times it felt normal however. Maybe we could improve UX a little here by adding some space between the search box and the letters? or you could even get fancy and make it slide down when the search box is selected and slide back up when it's not.
> 
> ![PXL_20220711_041006613 MP](https://user-images.githubusercontent.com/16514652/178191071-a45dc1d6-caaa-476a-94db-db75f63d0c71.jpg)

This has been updated with animation

cewert commented on components/search/SearchResults.brs:
```diff
@@ -1,3 +1,60 @@
+    end if
+    m.searchTask.observeField("results", "loadResults")
+    m.searchTask.query = query
+    m.top.overhangTitle = query
```
This adds the search query to the overhang but not the word "Search". Was there not enough room? Didn't like how it looked? Thoughts?

candry7731 replied:
I added this

candry7731 commented:
> > For the search page, can we add the search query to the overhang? And probably the word search too right? i.e. "Search" or "Search: searchString". The search query will help fill some of the empty space in general but would be most helpful for voice search since you are more likely to search for a longer string of text. We could make more room in the search box but that takes away room for search results. I think showing the full search query in the overhang is a nice compromise.
> 
> I will get this added to the overhang text! This would fall inline with the grid screen overhang text as the search string does show when searching in the grid screens.

This has been updated

neilsb commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -195,15 +240,15 @@ sub setMoviesOptions(options)
-        { "Title": tr("PLAY_COUNT"), "Name": "PlayCount" },
-        { "Title": tr("RELEASE_DATE"), "Name": "PremiereDate" },
-        { "Title": tr("RUNTIME"), "Name": "Runtime" }
+        { "Title": tr("Title"), "Name": "SortName" },
+        { "Title": tr("IMDB Rating"), "Name": "CommunityRating" },
+        { "Title": tr("Critic Rating"), "Name": "CriticRating" },
+        { "Title": tr("Date Added"), "Name": "DateCreated" },
+        { "Title": tr("Date Played"), "Name": "DatePlayed" },
+        { "Title": tr("Official Rating"), "Name": "OfficialRating" },
+        { "Title": tr("Play Count"), "Name": "PlayCount" },
+        { "Title": tr("Release Date"), "Name": "PremiereDate" },
+        { "Title": tr("Runtime"), "Name": "Runtime" }
```
Not sure if the reason these translation strings were updated was to make the consistent with the way most of the other translation strings were done, but the change as it stands means all these strings will need retranslated in all the other languages.

The way the inbuilt roku translation works, is that effectively the text passed into `tr()` is a key, and the `translation` text is returned.  While you've updated the _key_ in the en_US translation file, all the other language files will are now missing the "Critic Rating" (and others) key, and our volunteers will to re-enter this.   

I would suggest either updating the keys in all the other translation files, or (preferably) revert all these changes and put it back to "CRITIC_RATING" etc.

Translation in general is something I think we need to look at with respect to alternatives to the inbuild Roku functions.  
Using the inbuild stuff means we either use the English text as the key (which is the common way we do it), but this has the downside that if we want to change the text (e.g.  We decide that it would be better as "Crtici's Rating"), then either the key changes for all languages, or the key and value are very similar, but slightly different - which I think is confusing.

Alternatively we have a key like "CRITIC_RATING" that we can then tweak the desired text for English as often as we want without changing the key, but the downside to this is that if a language doesn't have a translation, the key is shown rather than it defaulting to English.  (We can probably fix this with some clever stuff in the build script however)

Either way, I think translation stuff needs looked at properly, but I don't want this PR to make each of these string require translated again into 17 odd different languages.

neilsb commented:
That's the translation stuff sorted. And voice search seems to be working well. 

I'm seeing a problem when selecting search from the options menu on the Home Screen. When the search page opens nothing seems to have focus, so I can't do anything. I'll try and have a look at it shortly.

jimdogx commented:
I'm getting a crash when selecting "*" and then "Search" from the Home Screen:

![Screen Shot 2022-08-24 at 5 43 07 PM](https://user-images.githubusercontent.com/18198517/186543034-5905f0b0-983a-477b-8b03-55e569461978.png)

candry7731 commented:
Jimi, working on this now


How do you have the Comment Details in your vscode?

[image: image.png]

On Wed, Aug 24, 2022 at 6:44 PM Jimi ***@***.***> wrote:

> I'm getting a crash when selecting "*" and then "Search" from the Home
> Screen:
>
> [image: Screen Shot 2022-08-24 at 5 43 07 PM]
> <https://user-images.githubusercontent.com/18198517/186543034-5905f0b0-983a-477b-8b03-55e569461978.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jellyfin/jellyfin-roku/pull/593#issuecomment-1226605770>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT6PDRAZH6R34QSFNIAMD3TV22XXDANCNFSM5VS2W6CQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Chris Andry

462 Browning Loop
Mandeville LA 70448
504-617-2924

jimdogx commented:
Crash is fixed.  However, if you click in to any library (Movies, TV Shows, etc), the word "Search" shows up at the top on top of the first thumbnail.

neilsb commented:
From the main search screen (accessed via Options menu) search results all look good, and seem to return expected results.

From the results, Movies, TV Shows and Episodes all seem to work as expected also.

However, choosing a `Person`, `Music Album` or `Music Track` item causes the app to crash, as it's trying to display them in the Movie screen, rather than going to appropriate page for the type.

candry7731 reviewed (commented):
Formatting

jimdogx commented on components/ItemGrid/ItemGrid.brs:
```diff
@@ -153,7 +194,13 @@ sub loadInitialItems()
+        else
+            m.loadItemsTask.recursive = false
+        end if
+        m.loadItemsTask.itemId = m.top.parentItem.parentFolder
```
This change breaks Collections for me.

jimdogx reviewed (changes requested):
Don't know if it matters, but if you speak your password (one character at a time), your password is shown.
